### PR TITLE
Merge/netcdf restart

### PR DESCRIPTION
**Pull request by greyNearing**
1. Enabled netCDF restart capabilities.
2. Fixed issues with output long names and units.
3. Consolidated the number of calls to read the hru/gru mapping from the Local Attributes file - reduced from 4 to 1.

### CODE REVIEW

**Comment by anewman89** [comment]
What happened on the first merge commit (b92f1b58601c88968938c16c1a948b1bb8d84669)?  I wouldn't have expected all these changes to come in at this point, or am I missing something?  The first two commits make sense for the restart, then it appears the merge introduced a bunch of syntax changes?

**Comment by greyNearing** [comment]
Sorry, I canceled the other PR because it contained an error.

All of the syntax changes (e.g., endif -> end if) I did so as to make all of the syntax consistent throughout the code.

**Comment by anewman89** [comment]
Ok, sounds good.  Just wanted to make sure it wasn't a merge issue or pull request with old syntax.

File: build/source/driver/multi_driver.f90 [hunk 1]
```diff
@@ -515,16 +483,16 @@ program multi_driver
 
 ! set default model parameters
 do iGRU=1,nGRU
- do localHRU=1,gru_struc(iGRU)%hruCount
+ do iHRU=1,gru_struc(iGRU)%hruCount
```
Comment (luckymichael) [inline build/source/driver/multi_driver.f90]: Are we changing back from `localHRU` to `iHRU`?

File: build/source/driver/multi_driver.f90 [hunk 2]
```diff
@@ -718,22 +701,22 @@ program multi_driver
...
+  do iHRU=1,gru_struc(iGRU)%hruCount
    
    ! read forcing data
    call read_force(&
                    ! input
                    modelTimeStep,                              & ! intent(in):    time step index
                    iGRU,                                       & ! intent(in):    index of gru
-                   localHRU,                                   & ! intent(in):    index of LOCAL hru
-                   gru_struc(iGRU)%hruInfo(localHRU)%hru_nc,   & ! intent(in):    index of hru in netcdf
+                   iHRU,                                   & ! intent(in):    index of LOCAL hru
```
Comment (luckymichael) [inline build/source/driver/multi_driver.f90]: align comments

File: build/source/engine/check_icond.f90 [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+ mixdform                      ! mixed form of Richards' equation
+implicit none
+private
+public::check_icond
+contains
+
+ ! ************************************************************************************************
+ ! public subroutine read_icond: read model initial conditions
+ ! ************************************************************************************************
+ subroutine check_icond(nGRU,nHRU,                     & ! number of GRUs and HRUs
```
Comment (luckymichael) [inline build/source/engine/check_icond.f90]: nHRU is not used.

File: build/source/engine/check_icond.f90 [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+     ! ***** snow
+     case(ix_snow)
+      ! (check liquid water)
+      if(mLayerVolFracLiq(iLayer) < 0._dp .or. mLayerVolFracLiq(iLayer) > 1._dp)then
+       write(message,'(a,1x,i0)') trim(message)//'cannot initialize the model with volumetric fraction of liquid water < 0 or > 1: layer = ',iLayer
+       err=20; return
+      end if
+      ! (check ice)
+      if(mLayerVolFracIce(iLayer) < 0.05_dp .or. mLayerVolFracIce(iLayer) > 0.80_dp)then
+       write(message,'(a,1x,i0)') trim(message)//'cannot initialize the model with volumetric fraction of ice < 0.05 or > 0.70: layer = ',iLayer
```
Comment (luckymichael) [inline build/source/engine/check_icond.f90]: > 0.8 ?

File: build/source/engine/check_icond.f90 [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+     scalarTheta = mLayerVolFracIce(iLayer)*(iden_ice/iden_water) + mLayerVolFracLiq(iLayer)
+    end if
+
+    ! check that the initial volumetric fraction of liquid water and ice is reasonable
+    select case(mlayerLayerType(iLayer))
+
+     ! ***** snow
+     case(ix_snow)
+      ! (check liquid water)
+      if(mLayerVolFracLiq(iLayer) < 0._dp .or. mLayerVolFracLiq(iLayer) > 1._dp)then
```
Comment (luckymichael) [inline build/source/engine/check_icond.f90]: Can the checks be separated like:
```
if (mLayerVolFracLiq(iLayer) < 0._dp) less than Zero error
if (mLayerVolFracLiq(iLayer) > 1._dp) larger than One error
```
And `mLayerVolFracLiq(iLayer)` will be checked with `theta_res` and `theta_sat` so is it a redundant check here for `[0, 1]`?

File: build/source/driver/multi_driver.f90 (hunk 2 above)
Comment (greyNearing) [inline build/source/driver/multi_driver.f90]: Fixed.

File: build/source/engine/check_icond.f90 (hunk 3 above)
Comment (greyNearing) [inline build/source/engine/check_icond.f90]: Fixed.

File: build/source/engine/check_icond.f90 (hunk 4 above)
Comment (greyNearing) [inline build/source/engine/check_icond.f90]: Changed error report to match threshold.

File: build/source/engine/check_icond.f90 (hunk 5 above)
Comment (greyNearing) [inline build/source/engine/check_icond.f90]: Fixed.

File: build/source/driver/multi_driver.f90 (hunk 1 above)
Comment (greyNearing) [inline build/source/driver/multi_driver.f90]: After the merge, half the lines in `multi_driver.f90` used iHUR and half used localHRU. I had to choose one, and I went with iHRU because it is intuitive, it matches the other loop syntax in the rest of the code, and iHRU specifically is used elsewhere in the code.

File: build/source/engine/read_attrb.f90 [hunk 6]
```diff
@@ -29,304 +29,328 @@ module read_attrb_module
  ! ************************************************************************************************
  ! public subroutine read_dimension: read HRU and GRU dimension information on local attributes
  ! ************************************************************************************************
- subroutine read_dimension(maxGRU,maxHRU,err,message)
+ subroutine read_dimension(attrFile,fileGRU,fileHRU,nGRU,nHRU,err,message,startGRU,checkHRU)
```
Comment (luckymichael) [inline build/source/engine/read_attrb.f90]: I think we had gone through discussion that we wanted to separate allocation and index assignment.

File: build/source/engine/read_attrb.f90 [hunk 7]
```diff
@@ -29,304 +29,328 @@ module read_attrb_module
...
 
+ ! **********************************************************************************************
+ ! (4) check that we have all the desired varaibles 
+ ! **********************************************************************************************
  ! check that we have all desired categorical variables
  if(any(.not.checkType))then
-  do iAtt=1,nVar_type
-   if(.not.checkType(iAtt))then; err=20; message=trim(message)//'missing variable ['//trim(type_meta(iAtt)%varname)//'] in local attributes file'; return; endif
+  do iCheck = 1,size(type_meta)
+   if(.not.checkType(iCheck))then; err=20; message=trim(message)//'missing variable ['//trim(type_meta(iCheck)%varname)//'] in lo    cal attributes file'; return; endif
```
Comment (luckymichael) [inline build/source/engine/read_attrb.f90]: typo `local`

File: build/source/engine/read_attrb.f90 [hunk 8]
```diff
@@ -29,304 +29,328 @@ module read_attrb_module
...
   end do
  endif
 
  ! check that we have all desired local attributes
  if(any(.not.checkAttr))then
-  do iAtt=1,nVar_attr
-   if(.not.checkAttr(iAtt))then; err=20; message=trim(message)//'missing variable ['//trim(attr_meta(iAtt)%varname)//'] in local attributes file'; return; endif
+  do iCheck = 1,size(attr_meta)
+   if(.not.checkAttr(iCheck))then; err=20; message=trim(message)//'missing variable ['//trim(attr_meta(iCheck)%varname)//'] in lo    cal attributes file'; return; endif
```
Comment (luckymichael) [inline build/source/engine/read_attrb.f90]: space in `local`

File: build/source/engine/read_attrb.f90 (hunk 6 above)
Comment (greyNearing) [inline build/source/engine/read_attrb.f90]: There are two things to notice here:

1. There is no way to separate the index assignment form the allocate without reading the file twice. Both allocating and assigning requires knowing the HRU/GRU mapping from the Attributes file, and if we do not store that information when we read it in for the allocation then we simply have to read it in again.

2. There is no generality to separating these allocation and assignment functions - these routines are both doing the same task, and cannot be used for any other task. The local and global allocate routines have some generality (although they could be made truly general by passing nSnow, nSoil, nStats, etc. rather than grabbing those from global variables). 

The bottom line is that there is nothing gained in terms of (a) functionality, (b) code readability or parsimony, or (c) run-time efficiency by repeating a file read.  Doing this reduces run-time efficiency, decreases code readability, and if we ever change the input structure, we have to change two redundant sections of code.

File: build/source/engine/read_attrb.f90 (hunk 8 above)
Comment (greyNearing) [inline build/source/engine/read_attrb.f90]: Fixed.

File: build/source/netcdf/def_output.f90 [hunk 9]
```diff
@@ -100,21 +100,21 @@ subroutine def_output(nHRU,nSoil,infile,err,message)
   fstring = adjustl(fstring)
   fname = trim(infile)//'_'//trim(fstring)//'.nc'
   call ini_create(nHRU,nSoil,trim(fname),ncid(iFreq),err,cmessage)
-  if(err/=0)then; message=trim(message)//trim(cmessage); return; endif
-  print "(A,A)",'Created output file:',trim(fname)
- enddo
+  if(err/=0)then; message=trim(message)//trim(cmessage); return; end if
+  print*,'Created output file:',trim(fname)
```
Comment (luckymichael) [inline build/source/netcdf/def_output.f90]: explicit format `'(A,A)'` can prevent unexpected newline in `fname` when `fname` is long.

File: build/source/netcdf/def_output.f90 [hunk 10]
```diff
@@ -317,18 +319,40 @@ subroutine def_variab(ncid,iFreq,hruDesire,timeDesire,metaData,ivtype,err,messag
...
    ! add parameter units
-   err = nf90_put_att(ncid,iVarId,'units',trim(metaData(iVar)%varunit))
+   catName = trim(metaData(iVar)%varunit) 
+   if (iStat==iLookStat%totl) then
+    ! make sure that the units of this varaible allow for integration
+    if ((index(catName,'s-1')<=0).and.(index(catName,'s-2')<=0).and.(index(catName,'W m-2')<=0)) then 
+     err=20
+     message=trim(message)//'trying to integrate a non-time variable: '//trim(metaData(iVar)%varName)//' - units: '//trim(catName)
+     return
+    endif
```
Comment (luckymichael) [inline build/source/netcdf/def_output.f90]: Since it is not affect the model run, probably just throw a warning message instead of raising an error and stopping the run.

File: build/source/netcdf/modelwrite.f90 [hunk 11]
```diff
@@ -321,16 +322,235 @@ subroutine writeTime(modelTimestep,outputTimestep,meta,dat,err,message)
...
+                         nHRU,             & ! intent(in): number of HRUs
+                         prog_meta,        & ! intent(in): prognostics metadata 
+                         prog_data,        & ! intent(in): prognostics data
+                         indx_meta,        & ! intent(in): index metadata 
+                         indx_data,        & ! intent(in): index data
+                         err,message)        ! intent(out): error control
+ ! --------------------------------------------------------------------------------------------------------
+ ! --------------------------------------------------------------------------------------------------------
+ ! access the derived types to define the data structures
+! USE data_types,only:hru_d                  ! length of substep
```
Comment (luckymichael) [inline build/source/netcdf/modelwrite.f90]: delete

File: build/source/engine/read_icond.f90 [hunk 12]
```diff
@@ -1,522 +0,0 @@
-! SUMMA - Structure for Unifying Multiple Modeling Alternatives
```
Comment (luckymichael) [inline build/source/engine/read_icond.f90]: Add the change to `Makefile`

File: build/source/engine/check_icond.f90 [hunk 13]
```diff
@@ -0,0 +1,277 @@
+! SUMMA - Structure for Unifying Multiple Modeling Alternatives
```
Comment (luckymichael) [inline build/source/engine/check_icond.f90]: add the file in `Makefile`

File: build/source/netcdf/def_output.f90 (hunk 10 above)
Comment (greyNearing) [inline build/source/netcdf/def_output.f90]: Better to have it stop before the run starts and tell the user that this is happening rather than to have it give the user unexpected behavior in the output file.

File: build/source/netcdf/def_output.f90 (hunk 9 above)
Comment (greyNearing) [inline build/source/netcdf/def_output.f90]: Fixed.

File: build/source/engine/read_icond.f90 (hunk 12 above)
Comment (greyNearing) [inline build/source/engine/read_icond.f90]: Fixed.

File: build/source/engine/check_icond.f90 (hunk 13 above)
Comment (greyNearing) [inline build/source/engine/check_icond.f90]: Fixed.

File: build/source/netcdf/modelwrite.f90 (hunk 11 above)
Comment (greyNearing) [inline build/source/netcdf/modelwrite.f90]: I'm not sure what to delete here?

Comment (bartnijssen) [inline build/source/netcdf/modelwrite.f90]: @luckymichael : Can you elaborate?

**Comment by bartnijssen** [comment]
@greyNearing: Please resolve conflict and let me know when this is ready to merge

File: build/source/netcdf/modelwrite.f90 (hunk 11 above)
Comment (luckymichael) [inline build/source/netcdf/modelwrite.f90]: comment line `! USE data_types,only:hru_d                  ! length of substep` can be deleted if it is not used.

**Comment by bartnijssen** [comment]
@greyNearing : This will resolve issue #125. Please update the branch and let me know when this is ready to merge.
